lib/utils.js

```diff
@@ -57,52 +59,138 @@ module.exports = class AccessUtils {
     const models = [ this.options.groupModel ].concat(this.getGroupContentModels());
 
     models.forEach(modelName => {
-      const Model = this.app.models[modelName];
-
-      if (typeof Model.observe === 'function') {
-        debug('Attaching access observer to %s', modelName);
-        Model.observe('access', (ctx, next) => {
-          const currentUser = this.getCurrentUser();
-
-          if (currentUser) {
-            // Do not filter if options.skipAccess has been set.
-            if (ctx.options.skipAccess) {
-              debug('skipAccess: true - skipping access filters');
-              return next();
-            }
+      this.attachAccessObserver(modelName);
+      this.attachBeforeSaveObserver(modelName);
+    });
+  }
 
-            // Do not filter if the request is being made against a single model instance.
-            if (_get(ctx.query, 'where.id')) {
-              debug('looking up by Id - skipping access filters');
-              return next();
-            }
 
-            // Do not apply filters if no group access acls were applied.
-            const loopbackContext = this.app.loopback.getCurrentContext();
-            const groupAccessApplied = loopbackContext && loopbackContext.get('groupAccessApplied') || false;
+  /**
+   * Add access observer to a given model
+   *
+   * @param {String} Model name to add hook to.
+   */
+  attachAccessObserver(modelName) {
+    const Model = this.app.models[modelName];
 
-            if (!groupAccessApplied) {
-              debug('acls not appled - skipping access filters');
-              return next();
-            }
+    if (!typeof Model.observe === 'function') {
+      return;
+    }
 
-            debug('%s observe access: query=%s, options=%o, hookState=%o',
-              Model.modelName, JSON.stringify(ctx.query, null, 4), ctx.options, ctx.hookState);
+    debug('Attaching access observer to %s', modelName);
+    Model.observe('access', (ctx, next) => {
+      const currentUser = this.getCurrentUser();
 
-            return this.buildFilter(currentUser.getId(), ctx.Model)
-              .then(filter => {
-                debug('original query: %o', JSON.stringify(ctx.query, null, 4));
-                const where = _isEmpty(ctx.query.where) ? filter : {
-                  and: [ ctx.query.where, filter ]
-                };
+      if (currentUser) {
+        // Do not filter if options.skipAccess has been set.
+        if (ctx.options.skipAccess) {
+          debug('skipAccess: true - skipping access filters');
+          return next();
+        }
 
-                ctx.query.where = where;
-                debug('modified query: %s', JSON.stringify(ctx.query, null, 4));
-              });
-          }
+        // Do not filter if the request is being made against a single model instance.
+        if (_get(ctx.query, 'where.id')) {
```

Comment:
Why not filter if the request is against just a single model instance?  If you need to break the firewall between companies then the request can include the skipAccess option.